File: src/main/java/com/apps/PortfolioApp/Controller/ProjectController.java
package com.apps.PortfolioApp.Controller;

import com.apps.PortfolioApp.DTO.ProjectDTO;
import com.apps.PortfolioApp.Service.ProjectService;
import jakarta.validation.Valid;
import org.springframework.web.bind.annotation.*;
import java.util.List;

@RestController
@RequestMapping("/projects")
public class ProjectController {

    //Step1 ----> Create an object from Project Service class :
    private final ProjectService projectService;

    //Step2 ----> Create a con:
    public ProjectController(ProjectService projectService) {
        this.projectService = projectService;
    }

    //step3 ----> create a function to save the project to db:
    @PostMapping
    public ProjectDTO insert(@RequestBody @Valid ProjectDTO projectDTO) {
        return this.projectService.insertProject(projectDTO);
    }

    //step4 ----> create a function to get all projects from db:
    @GetMapping
    public List<ProjectDTO> getAll() {
        return this.projectService.getAllProjects();
    }

    //step5 ----> create a function to get the Specific project by id from db:
    @GetMapping("/id/{id}")
    public ProjectDTO getById(@PathVariable int id) {
        return this.projectService.getProjectByID(id);
    }

    //step6 ----> create a function to get the Specific project by projectName from db:
    @GetMapping("/projectName/{name}")
    public ProjectDTO getByName(@PathVariable String name) {
        return this.projectService.getProjectByName(name);
    }

    //step7 ----> create a function to update the project to db:
    @PutMapping("/{name}")
    public ProjectDTO update(@PathVariable String name , @RequestBody @Valid ProjectDTO projectDTO){
        return this.projectService.updateProject(name, projectDTO);
    }

    //step8----> create a function to delete the Specific project by id from db:
    @DeleteMapping("/{id}")
    public void delete(@PathVariable int id) {
        this.projectService.deleteProject(id);
    }

}

File: lib/Controllers/project_controller.dart
import 'package:flutter/material.dart';
import 'package:protfolio_app/Models/project.dart';
import 'package:protfolio_app/services/project_service.dart';

class ProjectController extends ChangeNotifier {
  //step1 ----> here I will Create the class's variables :
  final ProjectService _service = ProjectService();
  List<Project> _projects = [];
  bool _isLoading = false;
  String? _errorMessage;

  //step2 ----> here I will Create a getter && setter  :
  List<Project> get projects => _projects;
  bool get isLoading => _isLoading;
  String? get errorMessage => _errorMessage;

  //step3 ----> here I will Create a function to load projects <get all projects> :
  Future<void> loadProjects() async {
    try {
      _isLoading = true;
      _errorMessage = null;
      notifyListeners();

      _projects = await _service.getAllProjects();
    } catch (exception) {
      _errorMessage = exception.toString();
    } finally {
      _isLoading = false;
      notifyListeners();
    }
  }

  //step4 ----> here I will Create a function to add project :
  Future<void> addProject(Project project) async {
    try {
      _isLoading = true;
      _errorMessage = null;

      notifyListeners();

      final newProject = await _service.addProject(project);
      _projects.add(newProject);
    } catch (e) {
      _errorMessage = e.toString();
    } finally {
      _isLoading = false;
      notifyListeners();
    }
  }

  //step5 ----> here I will Create a function to update project :
  Future<void> updateProject(String name, Project project) async {
    try {
      _isLoading = true;
      _errorMessage = null;

      notifyListeners();

      final updatedProject = await _service.updateProject(name, project);
      final index = _projects.indexWhere((s) => s.name == name);

      if (index != -1) {
        _projects[index] = updatedProject;
      }
    } catch (exception) {
      _errorMessage = exception.toString();
    } finally {
      _isLoading = false;
      notifyListeners();
    }
  }

  //step6 ----> here I will Create a function to delete project :
  Future<void> deleteProject(int id) async {
    try {
      _isLoading = true;
      _errorMessage = null;

      notifyListeners();

      await _service.deleteProject(id);

      _projects.removeWhere((project) => project.id == id);
    } catch (exception) {
      _errorMessage = exception.toString();
    } finally {
      _isLoading = false;
      notifyListeners();
    }
  }
}
